packages/http/src/router/matchBaseUrl.ts

```diff
@@ -1,23 +1,55 @@
+import * as E from 'fp-ts/lib/Either';
+import * as O from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+
 import { INodeVariable, IServer } from '@stoplight/types';
 import { MatchType } from './types';
 
 const variableRegexp = /{(.*?)}/g;
 
-export function matchBaseUrl(server: IServer, baseUrl: string): MatchType {
-  const templateMatchResult = matchRequestUrlToTemplateUrl(baseUrl, server.url, server.variables);
+export function matchBaseUrl(server: IServer, baseUrl: string) {
+  return pipe(
+    convertTemplateToRegExp(server.url, server.variables),
+    E.map(regex =>
+      pipe(
+        O.fromNullable(regex.exec(baseUrl)),
+        O.fold(
+          () => MatchType.NOMATCH,
+          matches => (matches.length > 1 ? MatchType.TEMPLATED : MatchType.CONCRETE)
+        )
+      )
+    ),
+    E.fold(
+      e => {
+        throw e;
+      },
+      matchType => matchType
+    )
+  );
+}
 
-  if (!templateMatchResult) {
-    return MatchType.NOMATCH;
-  }
+export function convertTemplateToRegExp(
+  urlTemplate: string,
+  variables?: { [name: string]: INodeVariable }
+): E.Either<Error, RegExp> {
+  return pipe(
+    O.fromNullable(variables),
+    O.fold(
+      () => E.right(urlTemplate),
+      vars => replaceString(vars, urlTemplate)
+    ),
+    E.map(regexString => new RegExp(`^${regexString}$`))
+  );
 
-  return templateMatchResult.length > 1 ? MatchType.TEMPLATED : MatchType.CONCRETE;
-}
+  function replaceString(vars: { [name: string]: INodeVariable }, input: string) {
+    return E.tryCatch<Error, string>(
+      () => replaceStringUnsafe(input),
+      e => (e instanceof Error ? e : new Error('unknown error'))
+    );
 
-export function convertTemplateToRegExp(urlTemplate: string, variables?: { [name: string]: INodeVariable }) {
-  const regexp = !variables
-    ? urlTemplate
-    : urlTemplate.replace(variableRegexp, (_match, variableName) => {
-        const variable = variables[variableName];
+    function replaceStringUnsafe(input: string): string {
+      return input.replace(variableRegexp, (_match, variableName) => {
+        const variable = vars[variableName];
         if (!variable) {
```

Comment:
I think this is the place where we should be using `E.tryCatch()` — if we can convert this into two regular `E.right` and `E.left` would even be a bonus point